testsuites/syncgateway/upgrade/testserver/test_upgrade.py

```diff
@@ -0,0 +1,424 @@
+
+    for doc in sg_docs:
+        key = doc["doc"]["_id"]
+        if terminator in key:
+            continue
+        rev = doc["doc"]["_rev"]
+        rev_gen = int(rev.split("-")[0])
+
+        try:
+            del doc["doc"]["_rev"]
+        except KeyError:
+            log_info("no _rev exists in the dict")
+
+        del doc["doc"]["_id"]
+        try:
+            del added_docs[key]["_id"]
+        except KeyError:
+            log_info("Ignoring id verification")
+        assert rev_gen == expected_doc_map[key], "revision mismatch"
+        assert len(doc["doc"]) == len(added_docs[key]), "doc length mismatch"
+
+    log_info("finished verify_sg_docs_revision_history.")
+
+
+def send_changes_termination_doc(db, cbl_db, terminator_doc_id, terminator_channel):
+    db.create_bulk_docs(number=1, id_prefix=terminator_doc_id, db=cbl_db, channels=terminator_channel)
+
+
+def update_docs(db, cbl_db, added_docs, doc_obj, terminator_doc_id_prefix):
+    log_info("Starting doc updates")
+    current_user_doc_ids = []
+    for doc in added_docs:
+        log_info(doc)
+        current_user_doc_ids.append(doc)
+
+    docs_per_update = 3
+    doc_revs = {}
+    terminator_doc_id = "{}_0".format(terminator_doc_id_prefix)
+    terminator_not_found_msg = "Termination doc not found"
+
+    while True:
+        log_info("randomly update docs waiting for terminator arrive...")
+        try:
+            term_doc = db.getDocument(cbl_db, terminator_doc_id)
+            if term_doc == -1:
+                log_info(terminator_not_found_msg)
+            else:
+                doc_type = term_doc.__class__.__name__
+                if doc_type == "MemoryPointer":
+                    log_info("update_docs: Found termination doc")
+                    log_info("update_docs: Updated {} docs".format(len(doc_revs)))
+                    return doc_revs
+                else:
+                    log_info("update_docs: doc object is retrieved correctly")
+        except Exception:
+            log_info(terminator_not_found_msg)
+
+        user_docs_subset_to_update = []
+        for _ in range(docs_per_update):
+            random_doc_id = random.choice(current_user_doc_ids)
+            user_docs_subset_to_update.append(random_doc_id)
+
+        cbl_db_docs_to_update = {}
+        for doc_id in user_docs_subset_to_update:
+            log_info("Updating doc_id: {}".format(doc_id))
+            doc_body = doc_obj.toMap(db.getDocument(cbl_db, doc_id))
+            # numOfUpdates counts how many updates are made to the current document
+            # starting index value from 2, 2 means numOfUpdates = 0, and 3 means numOfUpdates = 1
+            # the current framework take 0 as False and 1 as True, even though an integer type is expected
+            # this is temporary solution to avoid this issue
+            if "numOfUpdates" in doc_body:
+                doc_body["numOfUpdates"] += 1
+            else:
+                doc_body["numOfUpdates"] = 2 + 1
```

Comment:
it is supposed to be doc_body["numOfUpdates"] = 1 if key does not exist, you are assigning for first time to 1.

Reply:
when inserting the number of update as integer, it stored correctly in cbl, however, while reading it, our test framework treats 0 as false, 1 as true, then 2 or more as numbers.  as a solution for this case, we increased the starting index to avoid true/false type mismatch with numbers for easy comparison.